packages/repository-json-schema/test/unit/filter-json-schema.unit.ts

```diff
@@ -0,0 +1,158 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/repository-json-schema
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Entity, Filter, hasMany, model, property} from '@loopback/repository';
+import {expect} from '@loopback/testlab';
+import * as Ajv from 'ajv';
+import {getFilterJsonSchemaFor} from '../../src/filter-json-schema';
+import {JsonSchema} from '../../src';
+
+describe('getFilterJsonSchemaFor', () => {
+  @model()
+  class Order extends Entity {
+    @property({id: true})
+    id: number;
+
+    @property()
+    customerId: number;
+  }
+
+  @model()
+  class Customer extends Entity {
+    @property({id: true})
+    id: number;
+
+    @property()
+    name: string;
+
+    @hasMany(Order)
+    orders?: Order[];
+  }
+
+  let ajv: Ajv.Ajv;
+  let customerFilterSchema: JsonSchema;
+
+  beforeEach(() => {
+    ajv = new Ajv();
+    customerFilterSchema = getFilterJsonSchemaFor(Customer);
+  });
+
+  it('produces a valid schema', () => {
+    const isValid = ajv.validateSchema(customerFilterSchema);
+
+    const SUCCESS_MSG = 'Filter schema is a valid JSON Schema';
+    const result = isValid ? SUCCESS_MSG : ajv.errorsText(ajv.errors!);
+    expect(result).to.equal(SUCCESS_MSG);
+  });
+
+  it('allows an empty filter', () => {
+    expectSchemaToAllowFilter(customerFilterSchema, {});
+  });
+
+  it('allows all top-level filter properties', () => {
+    const filter: Required<Filter> = {
+      where: {id: 1},
+      fields: {id: true, name: true},
+      include: [{relation: 'orders'}],
+      offset: 0,
+      limit: 10,
+      order: ['id DESC'],
+      skip: 0,
+    };
+
+    expectSchemaToAllowFilter(customerFilterSchema, filter);
+  });
+
+  it('describes "where" as an object', () => {
+    const filter = {where: 'invalid-where'};
+    ajv.validate(customerFilterSchema, filter);
+    expect(ajv.errors || []).to.containDeep([
+      {
+        keyword: 'type',
+        dataPath: '.where',
+        message: 'should be object',
+      },
+    ]);
+  });
+
+  it('describes "fields" as an object', () => {
+    const filter = {fields: 'invalid-fields'};
+    ajv.validate(customerFilterSchema, filter);
+    expect(ajv.errors || []).to.containDeep([
```

Comment:
I'm curious; why are we doing `ajv.errors || []` here?

Reply:
When `ajv.validate` returns `true`, `ajv.errors` is set to `null`. As a result, `shouldjs` produced error message that seemed to me as more difficult to understand.

I am open to suggestions on how to improve this assertion and/or make the code easier to understand!